src/rules/prefer-repository-shorthand.ts

```diff
@@ -0,0 +1,99 @@
+import type ESTree from "estree";
+import type { AST as JsonAST } from "jsonc-eslint-parser";
+
+import { createRule } from "../createRule.js";
+
+const githubUrlRegex =
+	/^(?:git\+)?(?:ssh:\/\/git@|http?s:\/\/)?(?:www\.)?github\.com\//;
+
+const isGitHubUrl = (url: string) => githubUrlRegex.test(url);
+
+const cleanGitHubUrl = (url: string) =>
+	url.replace(githubUrlRegex, "").replace(/\.git$/, "");
+
+export default createRule({
+	create(context) {
+		return {
+			JSONProperty: (node) => {
+				if (
+					node.key.type === "JSONLiteral" &&
+					node.key.value === "repository" &&
+					node.parent.parent.parent.type === "Program"
+				) {
+					if (node.value.type === "JSONObjectExpression") {
+						const { properties } = node.value;
+						const typeProperty = properties.find(
+							(property) =>
+								property.key.type === "JSONLiteral" &&
+								property.key.value === "type",
+						);
+						const urlProperty = properties.find(
+							(property) =>
+								property.key.type === "JSONLiteral" &&
+								property.key.value === "url",
+						);
+						const directoryProperty = properties.find(
+							(property) =>
+								property.key.type === "JSONLiteral" &&
+								property.key.value === "directory",
+						);
+						if (
+							!directoryProperty &&
+							typeProperty?.value.type === "JSONLiteral" &&
+							typeProperty.value.value === "git" &&
+							urlProperty &&
+							urlProperty.value.type === "JSONLiteral" &&
+							typeof urlProperty.value.value === "string" &&
+							isGitHubUrl(urlProperty.value.value)
+						) {
+							context.report({
+								fix(fixer) {
+									return fixer.replaceText(
+										node.value as unknown as ESTree.Node,
+										JSON.stringify(
+											cleanGitHubUrl(
+												(
+													urlProperty.value as JsonAST.JSONLiteral
+												).value as string,
```

Comment:
[Style] These casts are only necessary because TypeScript loses type refinements (narrowed types) inside functions. We can get around this by putting the refined type in a variable:

```ts
const url = urlProperty.value;

context.report({
  // ...
});
```